File: rel/app/windows/Livebook.cs
using System;
using System.Diagnostics;
using System.Drawing;
using System.IO;
using System.Windows.Forms;

#nullable enable

namespace Livebook;

static class LivebookMain
{
    [STAThread]
    static void Main(string[] args)
    {
        var prefix = "open:";
        var url = "";

        if (args.Length == 1 && args[0].StartsWith(prefix))
        {
            var uri = new System.Uri(args[0].Remove(0, prefix.Length));
            url = uri.AbsoluteUri;
        }

        var logPath = getLogPath();

        if (ElixirKit.API.IsMainInstance("dev.livebook.Livebook"))
        {
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);
            Application.Run(new LivebookApp(url, logPath));

            var code = ElixirKit.API.Stop();
            if (code == 0) { return; }
            var message = $"Livebook exited with exit code {code}.\r\nLogs available at: {logPath}";
            MessageBox.Show(new Form() { TopMost = true }, message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
        }
        else
        {
            ElixirKit.API.Publish("open", url);
        }
    }

    internal static string getLogPath()
    {
        return Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "Livebook",
            "Logs",
            "Livebook.log"
        );
    }
}

class DummyForm : Form {}

class LivebookApp : ApplicationContext
{
    private NotifyIcon notifyIcon;
    private string? url;
    private string logPath;

    public LivebookApp(string url, string logPath)
    {
        this.logPath = logPath;

        ThreadExit += threadExit;

        ContextMenuStrip menu = new ContextMenuStrip();
        menu.Items.Add("Open", null, openClicked);
        menu.Items.Add("New Notebook", null, openNewNotebookClicked);
        menu.Items.Add(new ToolStripSeparator());

        var copyURLButton = menu.Items.Add("Copy URL", null, copyURLClicked);
        copyURLButton.Enabled = false;

        menu.Items.Add("View Logs", null, viewLogsClicked);
        menu.Items.Add("Open .livebookdesktop.bat", null, openBootScriptClicked);
        menu.Items.Add(new ToolStripSeparator());
        menu.Items.Add("Settings", null, openSettingsClicked);
        menu.Items.Add("Quit", null, quitClicked);
        notifyIcon = new NotifyIcon()
        {
            Text = "Livebook",
            Visible = true,
            Icon = Icon.ExtractAssociatedIcon(Application.ExecutablePath)!,
            ContextMenuStrip = menu
        };
        notifyIcon.Click += notifyIconClicked;

        Directory.CreateDirectory(Path.GetDirectoryName(logPath)!);

        ElixirKit.API.Start(
            name: "app",
            logPath: logPath,
            ready: () => {
                ElixirKit.API.Subscribe((name, data) =>
                {
                    switch (name)
                    {
                        case "url":
                            copyURLButton.Enabled = true;
                            this.url = data;
                            break;

                        default:
                            throw new Exception($"unknown event {name}");
                    }
                });

                ElixirKit.API.Publish("open", url);
            },
            exited: (exitCode) =>
            {
                Application.Exit();
            }
        );
    }

    private void threadExit(object? sender, EventArgs e)
    {
        notifyIcon.Visible = false;
    }

    private void notifyIconClicked(object? sender, EventArgs e)
    {
        MouseEventArgs mouseEventArgs = (MouseEventArgs)e;

        if (mouseEventArgs.Button == MouseButtons.Left)
        {
            ElixirKit.API.Publish("open", "");
        }
    }

    private void openClicked(object? sender, EventArgs e)
    {
        ElixirKit.API.Publish("open", "");
    }

    private void openNewNotebookClicked(object? sender, EventArgs e)
    {
        ElixirKit.API.Publish("open", "/new");
    }

    private void copyURLClicked(object? sender, EventArgs e)
    {
        System.Windows.Forms.Clipboard.SetText(url!);
    }

    private void viewLogsClicked(object? sender, EventArgs e)
    {
        Process.Start(logPath);
    }

    private void openBootScriptClicked(object? sender, EventArgs e)
    {
        var path =
            Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
                ".livebookdesktop.bat"
            );

        if (!File.Exists(path))
        {
            File.WriteAllText(path, @"@echo off
rem This file is used to configure Livebook before booting.
rem If you change this file, you must restart Livebook for your changes to take place.
rem See https://hexdocs.pm/livebook/readme.html#environment-variables for all available enviornment variables.

rem Allow Livebook to connect to remote machines over IPv6
rem set ERL_AFLAGS=-proto_dist inet6_tcp

rem Add directory to PATH
rem set PATH=C:\bin;%PATH%
");
        }

        Process.Start("notepad.exe", path);
    }

    private void openSettingsClicked(object? sender, EventArgs e)
    {
        ElixirKit.API.Publish("open", "/settings");
    }

    private void quitClicked(object? sender, EventArgs e)
    {
        notifyIcon.Visible = false;
        Application.Exit();
    }
}

File: rel/app/macos/Sources/Livebook/Livebook.swift
import AppKit
import ElixirKit

@main
public struct Livebook {
    public static func main() {
        let app = NSApplication.shared
        let delegate = AppDelegate()
        app.delegate = delegate
        app.run()
    }
}

class AppDelegate: NSObject, NSApplicationDelegate {
    private var statusItem: NSStatusItem!
    private var logPath: String!
    private var launchedByOpenURL = false
    private var initialURLs: [URL] = []
    private var url: String?
    private var bootScriptURL: URL!

    func applicationDidFinishLaunching(_ aNotification: Notification) {
        logPath = "\(NSHomeDirectory())/Library/Logs/Livebook.log"
        bootScriptURL = NSURL.fileURL(withPath: "\(NSHomeDirectory())/.livebookdesktop.sh")

        ElixirKit.API.start(
            name: "app",
            logPath: logPath,
            readyHandler: {
                if (self.initialURLs == []) {
                    ElixirKit.API.publish("open", "")
                }
                else {
                    for url in self.initialURLs {
                        ElixirKit.API.publish("open", url.absoluteString)
                    }
                }
            },
            terminationHandler: { process in
                if process.terminationStatus != 0 {
                    DispatchQueue.main.sync {
                        let alert = NSAlert()
                        alert.alertStyle = .critical
                        alert.messageText = "Livebook exited with error status \(process.terminationStatus)"
                        alert.addButton(withTitle: "Dismiss")
                        alert.addButton(withTitle: "View Logs")

                        switch alert.runModal() {
                        case .alertSecondButtonReturn:
                            self.viewLogs()
                        default:
                            ()
                        }
                    }
                }

                NSApp.terminate(nil)
            }
        )

        statusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.variableLength)
        let button = statusItem.button!
        let icon = NSImage(named: "Icon")!
        let resizedIcon = NSImage(size: NSSize(width: 18, height: 18), flipped: false) { (dstRect) -> Bool in
            icon.draw(in: dstRect)
            return true
        }
        button.image = resizedIcon
        let menu = NSMenu()

        let copyURLItem = NSMenuItem(title: "Copy URL", action: nil, keyEquivalent: "c")

        menu.items = [
            NSMenuItem(title: "Open", action: #selector(open), keyEquivalent: "o"),
            NSMenuItem(title: "New Notebook", action: #selector(openNewNotebook), keyEquivalent: "n"),
            .separator(),
            copyURLItem,
            NSMenuItem(title: "View Logs", action: #selector(viewLogs), keyEquivalent: "l"),
            NSMenuItem(title: "Open .livebookdesktop.sh", action: #selector(openBootScript), keyEquivalent: ""),
            .separator(),
            NSMenuItem(title: "Settings", action: #selector(openSettings), keyEquivalent: ","),
            NSMenuItem(title: "Quit", action: #selector(NSApplication.terminate(_:)), keyEquivalent: "q")
        ]
        statusItem.menu = menu

        ElixirKit.API.addObserver(queue: .main) { (name, data) in
            switch name {
            case "url":
                copyURLItem.action = #selector(self.copyURL)
                self.url = data
            default:
                fatalError("unknown event \(name)")
            }
        }
    }

    func applicationWillTerminate(_ aNotification: Notification) {
        ElixirKit.API.stop()
    }

    func applicationShouldHandleReopen(_ sender: NSApplication, hasVisibleWindows flag: Bool) -> Bool {
        ElixirKit.API.publish("open", "")
        return true
    }

    func application(_ app: NSApplication, open urls: [URL]) {
        if !ElixirKit.API.isRunning {
            initialURLs = urls
            return
        }

        for url in urls {
            ElixirKit.API.publish("open", url.absoluteString)
        }
    }

    @objc
    func open() {
        ElixirKit.API.publish("open", "")
    }

    @objc
    func openNewNotebook() {
        ElixirKit.API.publish("open", "/new")
    }

    @objc
    func copyURL() {
        let pasteboard = NSPasteboard.general
        pasteboard.clearContents()
        pasteboard.setData(url!.data(using: .utf8), forType: NSPasteboard.PasteboardType.URL)
    }

    @objc
    func viewLogs() {
        NSWorkspace.shared.open(NSURL.fileURL(withPath: logPath))
    }

    @objc
    func openSettings() {
        ElixirKit.API.publish("open", "/settings")
    }

    @objc
    func openBootScript() {
        let fm = FileManager.default

        if !fm.fileExists(atPath: bootScriptURL.path) {
            let data = """
            # This file is used to configure Livebook before booting.
            # If you change this file, you must restart Livebook for your changes to take place.
            # See https://hexdocs.pm/livebook/readme.html#environment-variables for all available environment variables.

            # # Allow Livebook to connect to remote machines over IPv6
            # export ERL_AFLAGS="-proto_dist inet6_tcp"

            # # Add Homebrew to PATH
            # export PATH=/opt/homebrew/bin:$PATH
            """.data(using: .utf8)

            fm.createFile(atPath: bootScriptURL.path, contents: data)
        }

        NSWorkspace.shared.open(bootScriptURL)
    }
}
